docs/source/garak.generators.base.rst

```diff
@@ -1,6 +1,47 @@
+* top_p - Optionally, a temperature param to pass to the underlying model.
+* active - Whether or not the class is active. Usually true, unless a generator is disabled for some particular reason.
+* generator_family_name - Generators each belong to a family, describing a group. This is often related to the module name - for example, ``openai.py`` contains classes for working with OpenAI models, whose generator_family_name is "openai".
+* context_len - The number of tokens in the model context window, or None
+* modality - A dictionary with two keys, "in" and "out", each holding a set of the modalities supported by the generator. "in" refers to prompt expectations, and "out" refers to output. For example, a text-to-text+image model would have modality: ``dict = {"in": {"text"}, "out": {"text", "image"}}``.
+* supports_multiple_generations - Whether or not the generator can natively return multiple outputs from a prompt in a single function call. When set to False, the ``generate()`` method will make repeated calls, one output at a time, until the requested number of generations (in ``generations``) is reached.
+
+Functions:
+
+#. **__init___()**: Class constructor. Call this from generators after doing local init. It does things like populating name variables, notifying generator startup to the user, and logging generator construction.
+
+#. **generate*()**: This method is mediating access to the underlying model or dialogue system. The ``generate()`` orchestrates all interaction with the dialogue service/model. It takes a prompt and, optionally, a number of output generations (``generations_this_call``). It returns a list of responses of length up to the number of output generations, with each member a prompt response (e.g. text). Since ``generate()`` involves a reasonable amount of logic, it is preferable to not override this function, and rather work with the hooks and sub-methods provided.
+
+The general flow in ``generate()`` is as follows:
+
+  #. Call the ``_pre_generate_hook()``.
+  #. Work out how many generations we're doing this call (if -1 is passed via ``generations_this_call``, the default count in self.generations is used).
+  #. If only one generation is requested, return the output of ``_call_model`` with 1 generation specified.
+  #. If the underlying model supports multiple generations, return the output of ``_call_model`` invoked with the full count of generations.
+  #. Otherwise, we need to assemble the outputs over multiple calls. There are two options here.
+    #. Is garak running with ``parallel_attempts > 1`` configured? In that case, start a multiprocessing pool with as many workers as the value of ``parallel_attempts``, and have each one of these work on building the required number of generations, in any order.
+    #. Otherwise, call ``_call_model()`` repeatedly to collect the requested number of generations.
+  #. Strip ``None`` responses from the outputs, and return the resulting list of prompt responses.
```

Comment:
This was a great time to ask this question. `None` is intended to indicate that generation failed, so would definitely be good for the situation you mention. Perhaps then `generate()` should not filter out `None`s at all. But also - perhaps when a model fails when given a blank prompt, it should return `None`, instead of exception handling and a string. I think that we're in a better place if any `str`-typed output from a generator is an indication that the model successfully output a string, and that this output is the string component of the output. Wdyt?

cf.: https://github.com/leondz/garak/pull/689#discussion_r1605561347